buckets/local/bucket.go

```diff
@@ -0,0 +1,297 @@
+	"github.com/ipfs/go-cid"
+	ds "github.com/ipfs/go-datastore"
+	syncds "github.com/ipfs/go-datastore/sync"
+	bstore "github.com/ipfs/go-ipfs-blockstore"
+	chunker "github.com/ipfs/go-ipfs-chunker"
+	offline "github.com/ipfs/go-ipfs-exchange-offline"
+	cbor "github.com/ipfs/go-ipld-cbor"
+	ipld "github.com/ipfs/go-ipld-format"
+	md "github.com/ipfs/go-merkledag"
+	"github.com/ipfs/go-merkledag/dagutils"
+	"github.com/ipfs/go-unixfs"
+	"github.com/ipfs/go-unixfs/importer/balanced"
+	"github.com/ipfs/go-unixfs/importer/helpers"
+	"github.com/ipfs/go-unixfs/importer/trickle"
+	options "github.com/ipfs/interface-go-ipfs-core/options"
+	"github.com/ipfs/interface-go-ipfs-core/path"
+	car "github.com/ipld/go-car"
+)
+
+func init() {
+	ipld.Register(cid.DagProtobuf, md.DecodeProtobufBlock)
+	ipld.Register(cid.Raw, md.DecodeRawBlock)
+	ipld.Register(cid.DagCBOR, cbor.DecodeBlock)
+}
+
+var (
+	// ignoredFilenames is a list of default ignored file names.
+	ignoredFilenames = []string{
+		".DS_Store",
+	}
+)
+
+const (
+	// archiveName is the name used for the car archive.
+	archiveName = ".textile/buck.car"
+
+	// PatchExt is used to ignore tmp files during a pull.
+	PatchExt = ".buckpatch"
+)
+
+// Bucket tracks a local bucket tree structure.
+type Bucket struct {
+	path   string
+	root   cid.Cid
+	tmp    ipld.DAGService
+	store  bstore.Blockstore
+	layout options.Layout
+}
+
+// NewBucket creates a new bucket with the given path.
+func NewBucket(pth string, layout options.Layout) (*Bucket, error) {
+	bs := bstore.NewBlockstore(syncds.MutexWrap(ds.NewMapDatastore()))
+	bsrv := bserv.New(bs, offline.Exchange(bs))
+	b := &Bucket{
+		path:   pth,
+		tmp:    md.NewDAGService(bsrv),
+		store:  bs,
+		layout: layout,
+	}
+	_ = b.load()
+	return b, nil
+}
+
+// load the car archive into the tmp blockstore.
+func (b *Bucket) load() error {
+	file, err := os.Open(filepath.Join(b.path, archiveName))
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	h, err := car.LoadCar(b.store, file)
+	if err != nil {
+		return err
+	}
+	if len(h.Roots) > 0 {
+		b.root = h.Roots[0]
+	}
+	return nil
+}
+
+// Path returns the current archive's root cid.
+func (b *Bucket) Path() path.Resolved {
+	return path.IpfsPath(b.root)
+}
+
+// Get returns the node at cid from the current archive.
+func (b *Bucket) Get(ctx context.Context, c cid.Cid) (ipld.Node, error) {
+	return b.tmp.Get(ctx, c)
+}
+
+// Archive creates an archive describing the current path.
+func (b *Bucket) Archive(ctx context.Context) error {
+	n, err := b.walkPath(ctx, b.path)
+	if err != nil {
+		return err
+	}
+	file, err := b.createArchive()
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	if err = car.WriteCarWithWalker(ctx, b.tmp, []cid.Cid{n.Cid()}, file, carWalker); err != nil {
+		return err
+	}
+	b.root = n.Cid()
+	return nil
+}
+
+func carWalker(n ipld.Node) ([]*ipld.Link, error) {
+	var links []*ipld.Link
+	for _, l := range n.Links() {
+		if l.Name != "" {
+			links = append(links, l)
+		}
```

Comment:
Yep, I looked for awhile into other ways to make that determination, but only came up with this. Doesn't feel ideal, but works in this case since you can't have a filename with named and empty string.

Reply:
Seems to me like the correct solution. I can't imagine other way.